src/dbnode/storage/series/buffer.go

```diff
@@ -89,53 +99,50 @@ type databaseBuffer interface {
 
 	Stats() bufferStats
 
-	// MinMax returns the minimum and maximum blockstarts for the buckets
-	// that are contained within the buffer. These ranges exclude buckets
-	// that have already been drained (as those buckets are no longer in use.)
-	MinMax() (time.Time, time.Time, error)
-
-	Tick() bufferTickResult
-
-	NeedsDrain() bool
+	Tick(versions map[xtime.UnixNano]BlockState) bufferTickResult
 
-	DrainAndReset() drainAndResetResult
-
-	Bootstrap(bl block.DatabaseBlock) error
+	Bootstrap(bl block.DatabaseBlock)
 
 	Reset(opts Options)
 }
 
 type bufferStats struct {
-	openBlocks  int
 	wiredBlocks int
 }
 
-type drainAndResetResult struct {
-	mergedOutOfOrderBlocks int
-}
-
 type bufferTickResult struct {
 	mergedOutOfOrderBlocks int
+	evictedBuckets         int
 }
 
 type dbBuffer struct {
-	opts              Options
-	nowFn             clock.NowFn
-	drainFn           databaseBufferDrainFn
-	pastMostBucketIdx int
-	buckets           [bucketsLen]dbBufferBucket
-	blockSize         time.Duration
-	bufferPast        time.Duration
-	bufferFuture      time.Duration
+	opts  Options
+	nowFn clock.NowFn
+
+	bucketsMap map[xtime.UnixNano]*BufferBucketVersions
+	// Cache of buckets to avoid map lookup of above.
+	bucketVersionsCache [bucketsCacheSize]*BufferBucketVersions
+	// This is an in order slice of the block starts in the bucketsMap.
+	// We maintain this to avoid sorting the map keys adhoc when we want to
+	// perform operations in chronological order.
+	inOrderBlockStarts []time.Time
```

Comment:
While buckets do have start times, those start times are referencing the starts of blocks, so I think we should be calling these what they are fundamentally referring to.

It would be convenient to use `time.Time` everywhere, since the helper functions are useful, but using `time.Time` in maps is not good practice since it contains timezone and other data, making equality comparison difficult.